System/Random.hs

```diff
@@ -439,137 +465,136 @@ runGenStateT g f = runStateT (f PureGenI) g
+-- `IO` and especially when dealing with exception handling and resource allocation, which is
+-- where `StateT` should never be used. For example writing a random number of bytes into a
+-- temporary file:
 --
 -- >>> import UnliftIO.Temporary (withSystemTempFile)
 -- >>> import Data.ByteString (hPutStr)
--- >>> let ioGen g = withSystemTempFile "foo.bin" $ \_ h -> uniformR (0, 100) g >>= flip uniformByteStringPrim g >>= hPutStr h
+-- >>> let ioGen g = withSystemTempFile "foo.bin" $ \_ h -> uniformR (0, 100) g >>= flip uniformByteString g >>= hPutStr h
 --
 -- and then run it:
 --
--- >>> runMutGenIO_ (mkStdGen 1729) ioGen
+-- >>> runGenM_ (IOGen (mkStdGen 1729)) ioGen
 --
-runMutGenIO :: (RandomGen g, MonadIO m) => g -> (MutGen g RealWorld -> m a) -> m (a, g)
-runMutGenIO g action = do
-  mutGen <- liftIO $ thawGen $ MutGen g
-  res <- action mutGen
-  MutGen g' <- liftIO $ freezeGen mutGen
-  pure (res, g')
-{-# INLINE runMutGenIO #-}
-
--- | Same as `runMutGenIO`, but discard the resulting generator.
-runMutGenIO_ :: (RandomGen g, MonadIO m) => g -> (MutGen g RealWorld -> m a) -> m a
-runMutGenIO_ g action = fst <$> runMutGenIO g action
-{-# INLINE runMutGenIO_ #-}
-
-
-newtype PrimGen g s = PrimGenI (MutableByteArray s)
-
-instance (s ~ PrimState m, PrimMonad m, RandomGen g, Prim g) =>
-         MonadRandom (PrimGen g) s m where
-  newtype Frozen (PrimGen g) = PrimGen g
-  thawGen (PrimGen g) = do
-    ma <- newByteArray (Primitive.sizeOf g)
-    writeByteArray ma 0 g
-    pure $ PrimGenI ma
-  freezeGen (PrimGenI ma) = PrimGen <$> readByteArray ma 0
-  uniformWord32R r = applyPrimGen (genWord32R r)
-  uniformWord64R r = applyPrimGen (genWord64R r)
-  uniformWord8 = applyPrimGen genWord8
-  uniformWord16 = applyPrimGen genWord16
-  uniformWord32 = applyPrimGen genWord32
-  uniformWord64 = applyPrimGen genWord64
-  uniformByteArray n = applyPrimGen (genByteArray n)
-
-applyPrimGen :: (Prim g, PrimMonad m, s ~ PrimState m) => (g -> (a, g)) -> PrimGen g s -> m a
-applyPrimGen f (PrimGenI ma) = do
-  g <- readByteArray ma 0
+-- @since 1.2
+newtype IOGen g s = IOGenI (IORef g)
+
+instance (RandomGen g, MonadIO m) => MonadRandom (IOGen g) RealWorld m where
+  newtype Frozen (IOGen g) = IOGen { unIOGen :: g }
+    deriving (Eq, Show, Read)
+  thawGen (IOGen g) = fmap IOGenI (liftIO $ newIORef g)
+  freezeGen (IOGenI gVar) = fmap IOGen (liftIO $ readIORef gVar)
+  uniformWord32R r = applyIOGen (genWord32R r)
+  {-# INLINE uniformWord32R #-}
+  uniformWord64R r = applyIOGen (genWord64R r)
+  {-# INLINE uniformWord64R #-}
+  uniformWord8 = applyIOGen genWord8
+  {-# INLINE uniformWord8 #-}
+  uniformWord16 = applyIOGen genWord16
+  {-# INLINE uniformWord16 #-}
+  uniformWord32 = applyIOGen genWord32
+  {-# INLINE uniformWord32 #-}
+  uniformWord64 = applyIOGen genWord64
+  {-# INLINE uniformWord64 #-}
+  uniformShortByteString n = applyIOGen (genShortByteString n)
+
+-- | Apply a pure operation to generator atomically.
```

Comment:
```suggestion
-- | Apply a pure operation to the generator.
```
I think this is copy pasta - AFAICT the `IORef` methods used here are not atomic.